In [2]:
import numpy as np
import pandas as pd
import matplotlib as mp
import csv
import os
import glob

In [3]:
# output = "C:\\Users\\canor\\Documents\\GitHub\\FL-Political-Analysis\\Florida Analysis\\FL 2014 by Precinct\\CSV Files"
# outPath = glob.glob("{0}\*.csv".format(output))

# folPath = "C:\\Users\\canor\\Documents\\GitHub\\FL-Political-Analysis\\Florida Analysis\\FL 2014 by Precinct\\Text Files"
# filPath = glob.glob("{0}\*.txt".format(folPath))

# # For future years, turn the below code into a function to be called on the directory of .txt or .csv files
# # def txt_to_csv(input_path,output_path,fields):

# for file in filPath:
#     with open(file, 'r', encoding='latin1') as input_file:
#         in_txt = csv.reader(input_file, delimiter='\t')
#         filename = os.path.splitext(os.path.basename(file))[0] + '.csv'
#         stripped = (line.strip() for line in input_file)
#         lines = (line.split("\t") for line in stripped if line)
#         with open(os.path.join(output, filename), 'w', encoding='utf-8') as output_file:
#             fields = ['county_code', 'county_name', 'elec_num', 'elec_date', 'elec_name', 'precinct_id', 'poll_loc', 'total_reg', 'total_reg_r', 'total_reg_d', 'total_reg_other', 'contest_name', 'district', 'contest_code', 'cand_or_issue', 'cand_party', 'cand_id', 'doe_num', 'vote_total']
#             writer = csv.writer(output_file, escapechar='\\')
#             writer.writerow(fields)
#             writer.writerows(lines)

In [4]:
# # merging the files
# joined_files = os.path.join(output, "*.csv")
  
# # A list of all joined files is returned
# joined_list = glob.glob(joined_files)
  
# # Finally, the files are joined
# df = pd.concat(map(pd.read_csv, joined_list), ignore_index=True)

## Step 1: Reformat and Ingest Data

In [5]:
# Combine county files into single dataframe
def merge_files(directory):
    target_files = glob.glob(directory)
    combined_df = pd.DataFrame()
    for file in target_files:
        df = pd.read_table(file, names=['county_code', 'county_name', 'elec_num', 'elec_date', 'elec_name', 'precinct_id', 'poll_loc', 'total_reg', 'total_reg_r', 'total_reg_d', 'total_reg_other', 'contest_name', 'district', 'contest_code', 'cand_or_issue', 'cand_party', 'cand_id', 'doe_num', 'vote_total'], encoding_errors='replace')
        combined_df = pd.concat([combined_df, df])
    return combined_df

In [6]:
location = "C:\\Users\\canor\\Documents\\GitHub\\FL-Political-Analysis\\Florida Analysis\\FL 2014 by Precinct\\*"

df = merge_files(location)

In [7]:
# Examine dataframe for size, dtypes, and NaNs
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 385323 entries, 0 to 1119
Data columns (total 19 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   county_code      385323 non-null  object 
 1   county_name      385323 non-null  object 
 2   elec_num         385323 non-null  int64  
 3   elec_date        385323 non-null  object 
 4   elec_name        385323 non-null  object 
 5   precinct_id      385323 non-null  object 
 6   poll_loc         340897 non-null  object 
 7   total_reg        385323 non-null  int64  
 8   total_reg_r      385323 non-null  int64  
 9   total_reg_d      385323 non-null  int64  
 10  total_reg_other  385323 non-null  int64  
 11  contest_name     385323 non-null  object 
 12  district         349577 non-null  object 
 13  contest_code     385323 non-null  int64  
 14  cand_or_issue    385323 non-null  object 
 15  cand_party       365969 non-null  object 
 16  cand_id          380165 non-null  float64
 17

In [8]:
df.head()

county_code county_name  elec_num   elec_date              elec_name  \
0         ALA     Alachua     10218  11/04/2014  2014 General Election   
1         ALA     Alachua     10218  11/04/2014  2014 General Election   
2         ALA     Alachua     10218  11/04/2014  2014 General Election   
3         ALA     Alachua     10218  11/04/2014  2014 General Election   
4         ALA     Alachua     10218  11/04/2014  2014 General Election   

  precinct_id                          poll_loc  total_reg  total_reg_r  \
0           1  01 First Baptist Church of Waldo       1352            0   
1           1  01 First Baptist Church of Waldo       1352            0   
2           1  01 First Baptist Church of Waldo       1352            0   
3           1  01 First Baptist Church of Waldo       1352            0   
4           1  01 First Baptist Church of Waldo       1352            0   

   total_reg_d  total_reg_other                      contest_name  \
0            0                0               U.S. Representative   
1            0                0               U.S. Representative   
2            0                0               U.S. Representative   
3            0                0               U.S. Representative   
4            0                0  Governor and Lieutenant Governor   

      district  contest_code              cand_or_issue cand_party  \
0   District 3        140030                   Ted Yoho        REP   
1   District 3        140030          Marihelen Wheeler        DEM   
2   District 3        140030  Howard Term Limits Lawson        NPA   
3   District 3        140030                 UnderVotes              
4                     160000      Scott / Lopez-Cantera        REP   

       cand_id  doe_num  vote_total  
0  100409333.0    60281         415  
1  100428206.0    61370         270  
2  108074441.0    61510          29  
3          0.0      902          12  
4  103093132.0    61253         419

## Step 2: Removing Duplicates

In [9]:
df.duplicated().sum()

0

No dupes! Great!

2a: Make sure all counties are accounted for

## Step 3: Removing Extraneous Columns

In [18]:
total_reg_r_attribute = df['total_reg_r'].unique()
print(total_reg_r_attribute)
print(len(total_reg_r_attribute))

[     0    491   1075    263    213   1242    722   1114    270    214
   1257    520   1052    743   1070    442    511    572    100   1013
    874    466    428    189    864   1372    728    529    261    915
   1376    451   1898    196    637   1292    394    776    309    493
    494   1024    746    283   1636    824    385    821   1067    783
    566   1034    689    275    455     37     26     63    831    443
    537    820    755     96    287    333    193    547    291    439
    241    841    361    167   1131     83    909    835    737    123
    795    981    334    421    837    337    410    717   1094    299
    388    541   1148    849    488    919    678    615   1536    738
   1810   1492   3039    585   1407   1186    628   1517   1263   1597
    923    893    980    653    777   1051    862    141    168    411
    318    959    237     85    321    223    632    811   1253    651
    586   1156    675    202    807   1083   1215    750    838   1462
   212

Looks like we have 168 unique values in this column of 385k rows... it's not nothing, but probably not enough data to be useful in our analysis in any way. Let's remove the column.

In [19]:
total_reg_d_attribute = df['total_reg_d'].unique()
print(total_reg_d_attribute)
print(len(total_reg_d_attribute))

[    0   625  1202   249   193  1359  1056  1054   267   553  1419   413
   615   742   948  1003   812   685  1899  1377  1214   360   652   397
   489  1120   767   459   291   520   884   529   957   205   879  1321
   443  1083   542   499  2299  1717  1165   347  1477  3404   581   424
   701  1332   289   668  2500   374   457   109     6    71   741  1263
   840   446   727   485   926   527   348   445   540   358   484   696
   833   416   244   907   682   909   628   947   552   703  1066   246
   659   787   345   273   461   776   475   387   749   751   285  1164
   939   386   435  1198   663  1546  1357  1952   426  1382  1053   603
  1226  1128  1432   784  1093  1311   738  1187  1346   943  2240  1650
   809   503   918   805   670   717   524   437   454   754   808   620
   550   836   557   351   613  1117   813   167   704   849   797  1035
   705   798  1313   972   852   970   478  1555  1492  1390  1284  1300
   651   796   215   693  1186  1072 10679 32159 33

Same deal here -- 172 unique values over 385k rows means this column is partial, at best. Let's remove it.

In [12]:
df['elec_num'].unique() # All values are 10218, not useful!

array([10218], dtype=int64)

In [20]:
df['elec_name'].unique() # All values are '2014 General Election', not useful!

array(['2014 General Election'], dtype=object)

In [21]:
df['total_reg_other'].unique() # There are unique values!

array([    0,   447,   707,   178,   107,   846,   589,   801,   168,
         217,   759,   276,   373,   518,   651,   519,   442,   371,
         264,   647,   724,   234,   396,   231,   321,   780,   485,
         295,   191,   398,   682,   318,   771,   147,   884,  1304,
         418,   807,   467,   488,   783,  1339,   684,   268,  1253,
        2289,   293,   399,   458,   742,   172,   556,  1770,   267,
         379,    62,     3,    68,    93,   471,   344,   243,   430,
         233,   167,   374,   244,   126,   218,   162,   420,   332,
         175,   319,   701,   123,   576,   389,   462,   105,   531,
         551,   174,   207,   539,   206,   157,   284,   582,   324,
         282,   269,   573,   727,  1099,   860,   403,   308,  1136,
         603,  1258,  1031,  1576,  1016,   827,   950,   890,  1022,
         542,   770,   703,   705,  1132,   977,   696,   428,   416,
         408,   367,   698,   351,   183,   365,   330,   438,   309,
         461,   648,

In [22]:
# How many unique values are there?
len(df['total_reg_other'].unique())

169

Not enough to really matter. Let's drop those columns:

In [23]:
df = df.drop(columns=['total_reg_r', 'total_reg_d', 'total_reg_other', 'elec_num', 'elec_name'])
df.head()

county_code county_name   elec_date precinct_id  \
0         ALA     Alachua  11/04/2014           1   
1         ALA     Alachua  11/04/2014           1   
2         ALA     Alachua  11/04/2014           1   
3         ALA     Alachua  11/04/2014           1   
4         ALA     Alachua  11/04/2014           1   

                           poll_loc  total_reg  \
0  01 First Baptist Church of Waldo       1352   
1  01 First Baptist Church of Waldo       1352   
2  01 First Baptist Church of Waldo       1352   
3  01 First Baptist Church of Waldo       1352   
4  01 First Baptist Church of Waldo       1352   

                       contest_name     district  contest_code  \
0               U.S. Representative   District 3        140030   
1               U.S. Representative   District 3        140030   
2               U.S. Representative   District 3        140030   
3               U.S. Representative   District 3        140030   
4  Governor and Lieutenant Governor                     160000   

               cand_or_issue cand_party      cand_id  doe_num  vote_total  
0                   Ted Yoho        REP  100409333.0    60281         415  
1          Marihelen Wheeler        DEM  100428206.0    61370         270  
2  Howard Term Limits Lawson        NPA  108074441.0    61510          29  
3                 UnderVotes                     0.0      902          12  
4      Scott / Lopez-Cantera        REP  103093132.0    61253         419

## Step 4: Dealing With NaN Cells

In [24]:
df_nan = df[df.isna().any(axis=1)]
df_nan

county_code county_name   elec_date precinct_id    poll_loc  total_reg  \
0            BRE     Brevard  11/04/2014         100         NaN       1914   
1            BRE     Brevard  11/04/2014         100         NaN       1914   
2            BRE     Brevard  11/04/2014         100         NaN       1914   
3            BRE     Brevard  11/04/2014         100         NaN       1914   
4            BRE     Brevard  11/04/2014         100         NaN       1914   
...          ...         ...         ...         ...         ...        ...   
7400         SEM    Seminole  11/04/2014           9  Precinct 9     266005   
7402         SEM    Seminole  11/04/2014           9  Precinct 9     266005   
7404         SEM    Seminole  11/04/2014           9  Precinct 9     266005   
7406         SEM    Seminole  11/04/2014           9  Precinct 9     266005   
7407         SEM    Seminole  11/04/2014           9  Precinct 9     266005   

             contest_name     district  contest_code          cand_or_issue  \
0     U.S. Representative   District 8        140080             Bill Posey   
1     U.S. Representative   District 8        140080      Gabriel Rothblatt   
2     U.S. Representative   District 8        140080           WriteinVotes   
3     U.S. Representative   District 8        140080              OverVotes   
4     U.S. Representative   District 8        140080             UnderVotes   
...                   ...          ...           ...                    ...   
7400             Governor          NaN        160000               Write-in   
7402         School Board   District 1        610001  Number of Under Votes   
7404         School Board   District 1        610001       Times Over Voted   
7406        State Senator  District 10        240100  Number of Under Votes   
7407        State Senator  District 10        240100       Times Over Voted   

     cand_party      cand_id  doe_num  vote_total  
0           REP  100928199.0    60309         730  
1           DEM  119321677.0    60345         337  
2                        0.0      900           2  
3                        0.0      901           0  
4                        0.0      902          58  
...         ...          ...      ...         ...  
7400         NP          NaN      900           4  
7402         NP          NaN      902           0  
7404         NP          NaN      901           0  
7406         NP          NaN      902           0  
7407         NP          NaN      901           0  

[83020 rows x 14 columns]

In [25]:
df_nan_cand_party = df[df['cand_party'].isna()]
df_nan_cand_party

county_code county_name   elec_date precinct_id  \
13932         PAL  Palm Beach   11/4/2014        1002   
13933         PAL  Palm Beach   11/4/2014        1004   
13934         PAL  Palm Beach   11/4/2014        1006   
13935         PAL  Palm Beach   11/4/2014        1008   
13936         PAL  Palm Beach   11/4/2014        1010   
...           ...         ...         ...         ...   
8677          POL        Polk  11/04/2014         410   
8678          POL        Polk  11/04/2014         412   
8679          POL        Polk  11/04/2014         419   
8680          POL        Polk  11/04/2014         422   
8681          POL        Polk  11/04/2014         423   

                                                poll_loc  total_reg  \
13932                                               1002          0   
13933                                               1004          2   
13934                                               1006          0   
13935                                               1008          0   
13936                                               1010          0   
...                                                  ...        ...   
8677    CALVARY FIRST ASSEMBLY OF GOD 4550 E JOHNSON AVE        898   
8678         POINCIANA COMMUNITY CENTER 395 MARIGOLD AVE       6517   
8679   TUSCANY PRESERVE AT LAKE MARION 1850 PACIFIC ROAD       4959   
8680   SUNLAKE TERRACE ESTATES 6555-16 OLD LAKE WILSO...        208   
8681            LAKE HAMILTON WOMAN'S CLUB 85 N OMAHA ST         35   

              contest_name     district  contest_code cand_or_issue  \
13932  JUDGE ALAN O. FORST          NaN        520401            NO   
13933  JUDGE ALAN O. FORST          NaN        520401            NO   
13934  JUDGE ALAN O. FORST          NaN        520401            NO   
13935  JUDGE ALAN O. FORST          NaN        520401            NO   
13936  JUDGE ALAN O. FORST          NaN        520401            NO   
...                    ...          ...           ...           ...   
8677             Senate 14  District 14        240140    UnderVotes   
8678             Senate 14  District 14        240140    UnderVotes   
8679             Senate 14  District 14        240140    UnderVotes   
8680             Senate 14  District 14        240140    UnderVotes   
8681             Senate 14  District 14        240140    UnderVotes   

      cand_party      cand_id  doe_num  vote_total  
13932        NaN  105859521.0       20           0  
13933        NaN  105859521.0       20           0  
13934        NaN  105859521.0       20           0  
13935        NaN  105859521.0       20           0  
13936        NaN  105859521.0       20           0  
...          ...          ...      ...         ...  
8677         NaN          0.0      902          38  
8678         NaN          0.0      902         113  
8679         NaN          0.0      902          94  
8680         NaN          0.0      902           2  
8681         NaN          0.0      902           8  

[19354 rows x 14 columns]

In [26]:
df_nan_cand_party['cand_or_issue'].unique()

array(['NO', 'YES', 'OverVotes', 'UnderVotes'], dtype=object)

Looks like rows containing NaNs in cand_party are either rows describing overvotes and undervotes, or pertain to nonpartisan races that are not relevant to this analysis. Let's double check that latter assumption, and then drop those rows:

In [31]:
df_nan_cand_party[~df_nan_cand_party['cand_or_issue'].isin(['UnderVotes', 'OverVotes'])]['contest_name'].unique()

array(['JUDGE ALAN O. FORST', 'JUDGE MARK W. KLINGENSMITH',
       'JUDGE W. MATTHEW STEVENSON', 'JUDGE MARTHA C. WARNER',
       'AMENDMENT 1', 'AMENDMENT 2', 'AMENDMENT 3',
       'COUNTY QUESTION - CHILDREN SERVICES',
       'COUNTY QUESTION - SCHOOL BOARD',
       'PALM BEACH GARDENS - QUESTION 1',
       'PALM BEACH GARDENS - QUESTION 2'], dtype=object)

Assumptions confirmed.

In [14]:
races = df['contest_name'].unique()
print(races)
print(len(races))

['U.S. Representative' 'Governor and Lieutenant Governor'
 'Attorney General' 'Chief Financial Officer'
 'Commissioner of Agriculture' 'Tax Collector' 'County Commissioner'
 'Retention of Robert T. Benton' 'Retention of Joseph Lewis, Jr.'
 'Retention of Scott Makar' 'Retention of Tim Osterhaus'
 'Retention of Clay Roberts'
 'Amendment No. 1: Water and Land Conservation - Dedicates funds to acquire and restore Florida conservation and recreation lands'
 'Amendment No. 2: Use of Marijuana for Certain Medical Conditions'
 'Amendment No. 3: PROSPECTIVE APPOINTMENT OF CERTAIN JUDICIAL VACANCIES'
 'State Representative' 'Clerk of the Circuit Court' 'School Board'
 'Canaveral Port Authority' 'Retention of Wendy W. Berger'
 'Retention of Kerry I. Evander' 'Retention of Charles Alan Lawson'
 'Retention of Richard B. Orfinger' 'Retention of William David Palmer'
 'Retention of Thomas D. Sawaya' 'Retention of F. Rand Wallis'
 'Circuit Judge' 'Barefoot Bay Recreation District (Vote for 3)'
 'Viera

In [15]:
def select_races(df):
    df = df[df.contest_name.isin(['Governor',\
                                'United States Senator',\
                                'U.S. REPRESENTATIVE'
                                'GOVERNOR AND  LT.GOVERNOR'
                                'STATE SENATOR' 
                                'STATE REPRESENTATIVE'
                                'Governor and Lieutenant Governor',\
                                'U.S. Representative',\
                                'State Senator',\
                                'State Representative'])]\
                                .reset_index()
    return df

In [16]:
df_narrowed = select_races(df)
df_narrowed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36551 entries, 0 to 36550
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   index            36551 non-null  int64  
 1   county_code      36551 non-null  object 
 2   county_name      36551 non-null  object 
 3   elec_num         36551 non-null  int64  
 4   elec_date        36551 non-null  object 
 5   elec_name        36551 non-null  object 
 6   precinct_id      36551 non-null  object 
 7   poll_loc         31601 non-null  object 
 8   total_reg        36551 non-null  int64  
 9   total_reg_r      36551 non-null  int64  
 10  total_reg_d      36551 non-null  int64  
 11  total_reg_other  36551 non-null  int64  
 12  contest_name     36551 non-null  object 
 13  district         34750 non-null  object 
 14  contest_code     36551 non-null  int64  
 15  cand_or_issue    36551 non-null  object 
 16  cand_party       36217 non-null  object 
 17  cand_id     

## Step (Final): Check for completeness
Finally, let's just double check to make sure all counties and races are present in our final output:

In [ ]:
counties = df['county_name'].unique()
print(counties)
print(len(counties))
# All counties are present and accounted for 